server/src/main/java/org/opensearch/tiering/HotToWarmTieringService.java

```diff
@@ -0,0 +1,874 @@
+        requestContext.addToNotFound(index.getName());
+        requestContext.removeFromInProgress(index);
+        requestUuidToRequestContext.put(reqId, requestContext);
+        indexTieringInfoMap.remove(index);
+    }
+
+    protected boolean isSearchNode(DiscoveryNode node) {
+        return node != null && node.isSearchNode();
+    }
+
+    protected ShardTieringStatus updateShardTieringStatus(final ShardRouting shard, ClusterState clusterState) {
+        final boolean isShardFoundOnSearchNode = isSearchNode(clusterState.getNodes().get(shard.currentNodeId()));
+        final boolean isShardRelocatingToSearchNode = isSearchNode(clusterState.getNodes().get(shard.relocatingNodeId()));
+        ShardTieringState tieringState;
+        String reason = null;
+        if (shard.started() && isShardFoundOnSearchNode) {
+            tieringState = ShardTieringState.SUCCESSFUL;
+        } else if (shard.unassigned()) {
+            tieringState = ShardTieringState.INIT;
+            reason = "Shard is unassigned due to " + shard.unassignedInfo().getReason();
+        } else if (((shard.initializing() || shard.started()) && !isShardFoundOnSearchNode)
+            || (shard.relocating() && !isShardRelocatingToSearchNode)) {
+                tieringState = ShardTieringState.FAILED;
+                reason = "Shard with current state: "
+                    + shard.state().toString()
+                    + " is neither allocated nor relocating to the search node, "
+                    + "current node: "
+                    + shard.currentNodeId()
+                    + ", relocating node: "
+                    + shard.relocatingNodeId();
+            } else {
+                tieringState = ShardTieringState.PROCESSING;
+            }
+        return new ShardTieringStatus(
+            shard.relocating() ? shard.relocatingNodeId() : shard.currentNodeId(),
+            shard.state(),
+            tieringState,
+            reason
+        );
+    }
+
+    private void updateIndexShardStatus(ClusterState clusterState) {
+        List<ShardRouting> shardRoutings;
+        // to include successful/failed indices
+        final Set<IndexTieringInfo> completedIndices = new HashSet<>();
+        for (Index index : indexTieringInfoMap.keySet()) {
+            IndexTieringInfo indexTieringInfo = indexTieringInfoMap.get(index);
+            if (indexTieringInfo.isCompleted()) {
+                continue;
+            }
+            try {
+                // Ensure index is not deleted
+                shardRoutings = clusterState.routingTable().allShards(index.getName());
+            } catch (IndexNotFoundException ex) {
+                // Index already deleted nothing to do
+                logger.warn("Index [{}] deleted before hot to warm relocation finished", index.getName());
+                processNotFoundIndex(indexTieringInfo);
+                continue;
+            }
```

Comment:
This probably isn't unique to the tiering changes but it would be great to have a helper method in the `clusterState.routingTable()` to check if an index is present or not instead of relying on a try/catch.

Reply:
updated the logic here